core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -495,44 +498,37 @@ class GroupCoordinator(val brokerId: Int,
         responseCallback(Errors.UNKNOWN_MEMBER_ID)
 
       case Some(group) => group.inLock {
-        group.currentState match {
-          case Dead =>
-            // if the group is marked as dead, it means some other thread has just removed the group
-            // from the coordinator metadata; it is likely that the group has migrated to some other
-            // coordinator OR the group is in a transient unstable phase. Let the member retry
-            // joining without the specified member id.
-            responseCallback(Errors.UNKNOWN_MEMBER_ID)
-
-          case Empty =>
-            responseCallback(Errors.UNKNOWN_MEMBER_ID)
-
-          case CompletingRebalance =>
-            if (!group.has(memberId))
+        if (!group.validateMemberIdIfStatic(memberId, groupInstanceId)) {
+          responseCallback(Errors.FENCED_INSTANCE_ID)
+        } else if (!group.has(memberId)) {
+          responseCallback(Errors.UNKNOWN_MEMBER_ID)
+        } else if (generationId != group.generationId) {
+          responseCallback(Errors.ILLEGAL_GENERATION)
+        } else {
+          group.currentState match {
+            case Dead =>
+              // if the group is marked as dead, it means some other thread has just removed the group
+              // from the coordinator metadata; it is likely that the group has migrated to some other
+              // coordinator OR the group is in a transient unstable phase. Let the member retry
+              // joining without the specified member id.
```

Comment:
Static membership makes this handling a little dangerous. If the consumer resets its memberId to unknown, then we could end up fencing a newer member. I wonder if it would be safer to use COORDINATOR_NOT_AVAILABLE and let the consumer lookup the coordinator again and retry with the existing memberId.

Reply:
Not sure I'm following here, does it help by moving dead group logic (UNKNOWN_MEMBER_ID) to the very top?

Reply:
I don't think it helps. The point is that this is a spurious UNKNOWN_MEMBER_ID. For dynamic members, it doesn't matter because there is no harm resetting the member id. For static members, resetting the memberId means we will fence the active member (which may be a different instance).

cc @guozhangwang What do you think?

Reply:
I think this goes back to our discussion in https://github.com/apache/kafka/pull/6177#discussion_r250660764
We could not completely eliminate the risk of resetting a culprit consumer generation if we don't restrict `resetGeneration` usage. However, when getting fenced exception, user should understand there is some severe configuration error that needs to be addressed. So I feel as long as we could fence and indicate the problem, user should be able to handle that.

Reply:
Is there a downside to using COORDINATOR_NOT_AVAILABLE as the error instead? Wouldn't it be nice not to have unnecessary resetting of the memberId?

Reply:
I couldn't think of any. We could use it while group is dead. @guozhangwang Do you see any downsides?